### ISSUE TITLE
Read std::vector

### ISSUE THREAD

**Issue by abbradar** [danielaparker/jsoncons#97]
Hi,

Previously I was able to deserialize `std::vector` with just

```
auto a = json.template as<vector<foo>>();
```

Nowadays on `master` this gives me the following error:

```
ext/jsoncons/src/jsoncons/json.hpp: In instantiation of 'T jsoncons::basic_json<CharT, JsonTraits, Allocator>::as() const [with T = std::vector<instance_settings>; CharT = char; JsonTraits = jsoncons::json_traits<char>; Allocator = std::allocator<char>]':
include/cgate_settings.h:87:21:   required from 'static router_settings jsoncons::json_type_traits<J, router_settings>::as(const J&) [with J = jsoncons::basic_json<char, jsoncons::json_traits<char>, std::allocator<char> >]'
ext/jsoncons/src/jsoncons/json.hpp:2230:49:   required from 'T jsoncons::basic_json<CharT, JsonTraits, Allocator>::as() const [with T = router_settings; CharT = char; JsonTraits = jsoncons::json_traits<char>; Allocator = std::allocator<char>]'
src/dummy.cpp:175:51:   required from here
ext/jsoncons/src/jsoncons/json.hpp:2230:49: error: 'as' is not a member of 'jsoncons::json_type_traits<jsoncons::basic_json<char, jsoncons::json_traits<char>, std::allocator<char> >, std::vector<instance_settings>, void>'
         return json_type_traits<json_type,T>::as(*this);
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
```

What's going on here? As it builds with an old revision of jsoncons I expect that maybe some other method needs to be used now?

Thanks for your help!

**Comment by danielaparker** [comment]
I don't see how that could have worked for an arbitrary type `foo`, does `router_settings` have a constructor that takes a `json` parameter? Could you post enough of `router_settings` so I can use it for a test case?

Thanks,
Daniel

**Comment by danielaparker** [comment]
You can serialize/deserialize to/from your own types by extending `json_type_traits` in the `jsoncons` namespace, e.g.

```c++
struct settings
{
    std::string name;
    double value;
};

namespace jsoncons
{
    template<class Json>
    struct json_type_traits<Json, settings>
    {
        static const bool is_assignable = true;

        static bool is(const Json& rhs) noexcept
        {
            return rhs.count("name") == 1 && rhs.count("value") == 1  && rhs.is_object();
        }
        static settings as(const Json& rhs)
        {
            settings val;
            val.name = rhs["name"].as<std::string>();
            val.value = rhs["value"].as<double>();
            return val;
        }
        static Json to_json(settings val)
        {
            Json j;
            j["name"] = val.name;
            j["value"] = val.value;
            return j;
        }
    };

};

int main()
{
    settings a{ "a",1.0 };

    json j = a;

    std::cout << "(1) " << std::boolalpha << j.is<settings>() << "\n\n";

    std::cout << "(2) " << j << "\n\n";

    auto b = j.as<settings>();

    std::cout << "(3) " << b.name << " = " << b.value << "\n\n";

    std::vector<settings> settings_in{ settings{ "a",1.0 }, settings{ "b",2.0 } };

    json ja = settings_in;

    std::cout << "(4) " << std::boolalpha << ja.is<std::vector<settings>>() << "\n\n";

    std::cout << "(5)" << "\n";
    std::cout << ja << "\n\n";

    auto settings_out = ja.as<std::list<settings>>();

    std::cout << "(6)" << std::endl;
    for (auto setting : settings_out)
    {
        std::cout << setting.name << " = " << setting.value << std::endl;
    }
}
```
The output is
```
(1) true

(2) {"name":"a","value":1.0}

(3) a = 1

(4) true

(5)
[{"name":"a","value":1.0},{"name":"b","value":2.0}]

(6)
a = 1
b = 2
```

**Comment by abbradar** [comment]
This test compiles successfully on ba96033, but not on master:

```
#include <jsoncons/json.hpp>

using namespace std;
using namespace jsoncons;

struct foo {
  string test_str;
};

namespace jsoncons {
  template<class J> struct json_type_traits<J, foo> {
    static foo as(const J& json) {
      foo res;
      res.test_str = json["test_str"].template as<string>();
      return res;
    }
  };
}

int main()
{
  json test;
  test.template as<vector<foo>>();
}
```

**Comment by abbradar** [comment]
Aha, it looks like I need `static const bool is_assignable = true;` for this to work! What does this constant mean?

**Comment by abbradar** [comment]
It's also interesting that this constant is not required when deserializing just `settings`, only when it's in a container.

**Comment by danielaparker** [comment]
Thanks for your comments.

When `json_type_traits` was generalized to handle all standard library containers, it was necessary to constrain the value type and mapped type for the sequence and associative containers to types that also had specializations in `json_type_traits`. To allow this, I added the `is_assignable` data member to distinguish between types that did and did not have specializations. However, after thinking about your comments, I came up with an alternative implementation that does not require the `is_assignable` data member (it only requires something in the default template). I've updated master accordingly, so the example you posted should now work without modification.
